test/Test/Pos/Ssc/GodTossing/Toss/BaseSpec.hs

```diff
@@ -0,0 +1,573 @@
+        -- shares distribution.
+        someRichman =
+            case res of
+                Just s -> s
+                _ -> error "Test.Pos.Ssc.GodTossing.Toss.BaseSpec.checkBadCommsPayload:\
+                           \ empty commitments payload"
+          where holders = HM.keys . getCommitmentsMap $ commsMap
+                randomIndex = fst $ randomR (0, length holders - 1) $ mkStdGen seed
+                res = holders ^? ix randomIndex
+
+        gtgs' = gtgs & gsCommitments %~ wrapCMap (HM.insert someRichman comm)
+        commitmentAlreadySent =
+            tossRunner mrs gtgs' $ checkCommitmentsPayload epoch commsMap
+        res3 = case commitmentAlreadySent of
+            Left (CommitmentAlreadySent (s :| [])) -> s == someRichman
+            _ -> False
+
+    in not (commMember sid _gsCommitments) &&
+       not (commMember sid commsMap) &&
+       not (null _gsCommitments) &&
+       not (null commsMap) ==> res1  && res2 && res3
+
+newtype GoodOpeningPayload = GoodOpens
+    { getGoodOpens :: (GtGlobalState, OpeningsMap)
+    } deriving (Show, Eq)
+
+instance Arbitrary GoodOpeningPayload where
+    arbitrary = GoodOpens <$> do
+
+      -- These fields won't be used, so they can be entirely arbitrary
+        _gsShares <- arbitrary :: Gen SharesMap
+        _gsVssCertificates <- arbitrary :: Gen VssCertData
+
+        -- Because 'Opening's and 'Commitment's in the openings payload map and
+        -- '_gsCommitments' resp. will need to be matched succesfully by
+        -- 'checkOpeningsPayload', they must be generated in tandem, along with their
+        -- respective keys in both maps and the dummy 'CommitmentSignature' to be used in
+        -- the commitments map.
+        commsAndOpens <- arbitrary
+            :: Gen [(StakeholderId, (PublicKey, CommitmentOpening, CommitmentSignature))]
+        let fun (s, (pk, p, cs)) = (s, ((pk, coCommitment p, cs), coOpening p))
+
+            -- The stakeholders in this assoc list are paired with a tuple, the first
+            -- component of which will be their value in '_gsCommitments' and the second
+            -- the value in '_gsOpenings'.
+            stakeHoldersAndCOs
+                :: [(StakeholderId,
+                    ((PublicKey, Commitment, CommitmentSignature), Opening))]
+            stakeHoldersAndCOs = map fun commsAndOpens
+            _gsCommitments =
+                mkCommitmentsMapUnsafe .
+                -- over _2 (view _1) (a,(b,c)) = (a,b)
+                HM.fromList $ fmap (over _2 $ view _1) stakeHoldersAndCOs
+        openingPldList <- sublistOf stakeHoldersAndCOs
+
+        -- For the test data to be correct, none of the stakeholders with an opening in
+        -- openings payload can have an opening in the global state i.e. be a key in
+        -- '_gsOpenings'.
+        _gsOpenings <- customHashMapGen
+            (arbitrary `suchThat` (not . flip elem (map fst openingPldList)))
```

Comment:
I thought about that, but then I remembered what you told me almost a year ago when I wanted to use `IntSet`s instead of regular `Set Int`s for some trivial task: premature optimization is the root of all evil.

Don't get me wrong, you're right, `elem` is bad when compared to `O(log n)` lookup, especially since this hashmap may be quite large, which implies a large number of lookups to this possibly large list too.

I decided that I'd just do it this way quickly and move on with the issue (it was already overdue for several weeks) than to ponder for half an hour whether or not the `O(n * log n)` conversion would slow down tests enough to hang CI.